impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,196 @@
+\end{itemize}
+
+
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_impute_constant(constant : DA::NonNull):
+    # instead of VectorDomain(DA), we add ::new() to get a new instance of DA. This is bcause DA has the ImputableDomain trait.
+    input_domain = VectorDomain(DA::new())
+    output_domain = VectorDomain(AllDomain(DA::NonNull))
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    assert(not constant.is_null); # not DA::is_null(constant) |\label{line:null}|
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
+        def impute_constant(x: DA) -> DA::NonNull: |\label{line:atom-fn}|
+            return constant if x.is_null else x
+        return list(map(impute_constant, data)) |\label{line:map}|
+        
+    # can we comment out input_metric, output_metric, and stability_relation, and function if it's not being called anymore?
+    return make_row_by_row(input_domain, output_domain, impute_constant); |\label{line:row-transform}|
+\end{lstlisting}
+
+\subsection{Proof}
+
+\begin{lemma}[\texttt{DA::NonNull} contains \texttt{null}] \label{lemma:NonNull}
+A \texttt{var} of type \texttt{DA::NonNull} can be of type \texttt{null}.
+\end{lemma}
+\begin{proof}
+Let the domain of atom variable \texttt{DA} be \texttt{InherentNullDomain<AllDomain<f64>>}. Recall that \texttt{InherentNullDomain} exists for types that can represent null inherently in the carrier type. Then the type $$\texttt{DA::NonNull == InherentNullDomain<AllDomain<f64>>::NonNull == f64}.$$ The latter holds because in the \texttt{InherentNullDomain} implementation in the rust code \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/impute.rs#L48-L56}, the \texttt{type NonNull = Self::Carrier}. The \texttt{Carrier} of \texttt{VectorDomain<AllDomain<T>>} has type \texttt{T}, so in this case the \texttt{::Carrier} is type \texttt{f64}.
+
+Therefore \texttt{var} is also of type \texttt{f64}. \texttt{f64} can contain \texttt{null} values, so we are done.
+\end{proof}
+
+\begin{lemma}[Precondition for row transform]\label{lemma:pure-fn}
+\texttt{make\_impute\_constant} satisfies the preconditions of \texttt{make\_row\_by\_row}. That is, the atom function \texttt{impute\_constant} in pseudo code line \ref{line:atom-fn} is a pure function. 
+\end{lemma}
+\begin{proof}
+We reference the examples in \href{https://en.wikipedia.org/wiki/Pure_function}{Pure function Wikipedia page} for the definition of a pure function, and later as a check list to verify the properties of a pure function hold. To verify whether \texttt{impute\_constant} is a pure function, it must satisfy the following properties:
```

Comment:
we should not refer to the Wikipedia page, but rather include the definition of pure function we are using in one of our definitions documents

Reply:
added in def 4.3